### PR TITLE
Move zbuf.Record into package zng etc

### PR DESCRIPTION
**Pull request by mccanne**
This commit recreates the zng.Record value type based on the
zbuf.Record and deletes zbuf.Record.  It also merges zbuf.Descriptor
into zng.TypeRecord and eliminates the former.  This gets us
ready to create the generic type code and type context concepts for
the updated ZNG spec.

The zng writer also outputs in "canonical form", where each record
descriptor starts at id 0 and increases independent of the type IDs
in the internal stream.

resolver.Table is renamed resolver.Context to reflect its role as
the type context in the new ZNG spec and names throught are changed
to reflext "type context" over "descriptor table".

Unmarshaling of records has been updated to deal with the need to
create records and zng.Columns with a type context.  This is the
first move and will be imrpoved in a subsequent commit.

zbuf errors that were really zng errors have been moved to package zng.

### CODE REVIEW

File: zng/record.go [hunk 1]
```diff
@@ -2,26 +2,50 @@ package zng
...
+		LUT:     r.LUT,
+		TsCol:   r.TsCol,
+		Key:     r.Key,
+	}
 }
 
-func recordString(columns []Column) string {
-	return columnList("record[", columns, "]")
+//XXX this should take ID as arg
+func NewTypeRecord(id int, columns []Column) *TypeRecord {
```
Comment (mattnibs) [inline zng/record.go]: I this comment still relevant? id isn't used.

File: zng/record.go [hunk 2]
```diff
@@ -2,26 +2,50 @@ package zng
...
+		Columns: columns,
+		TsCol:   -1,
+		Key:     ColumnString("", columns, ""), //XXX
+	}
+	r.createLUT()
+	return r
+}
+
+//XXX
+func RecordString(columns []Column) string {
```
Comment (mattnibs) [inline zng/record.go]: Shouldn't this be called `RecordTypeString`?

File: zng/record.go [hunk 3]
```diff
@@ -2,26 +2,50 @@ package zng
...
+	return ColumnString("record[", columns, "]")
 }
 
 func (t *TypeRecord) String() string {
-	return recordString(t.Columns)
+	// XXX this needs review
+	//return ColumnString("", t.Columns, "")
+	return ColumnString("record[", t.Columns, "]")
```
Comment (mattnibs) [inline zng/record.go]: This looks fine, get rid of comment.

File: zng/record.go [hunk 4]
```diff
@@ -32,68 +56,15 @@ func (t *TypeRecord) UnmarshalJSON(data []byte) error {
...
-			in = rest
-			continue
-		}
-		rest, ok = match(rest, "]")
-		if ok {
-			return rest, LookupTypeRecord(columns), nil
-		}
-		return "", nil, ErrTypeSyntax
-	}
+// FinishUnmarsal is called after the Column types have been filled in
```
Comment (mattnibs) [inline zng/record.go]:
```suggestion
// FinishUnmarshal is called after the Column types have been filled in
```

File: zng/record.go [hunk 5]
```diff
@@ -32,68 +56,15 @@ func (t *TypeRecord) UnmarshalJSON(data []byte) error {
...
+// by the type resolver.
+func (t *TypeRecord) FinishUnmarsal(data []byte) {
```
Comment (mattnibs) [inline zng/record.go]:
```suggestion
func (t *TypeRecord) FinishUnmarshal(data []byte) {
```

Comment (mattnibs) [reply zng/record.go]: Also doesn't look like this function is called anywhere.

File: zng/record.go (hunk 1 above)
Comment (mccanne) [reply zng/record.go]: +1

File: zng/record.go (hunk 2 above)
Comment (mccanne) [reply zng/record.go]: How about` TyperRecordString`

File: zng/record.go (hunk 3 above)
Comment (mccanne) [reply zng/record.go]: +1

File: zng/record.go (hunk 4 above)
Comment (mccanne) [reply zng/record.go]: +1

File: zng/record.go (hunk 5 above)
Comment (mccanne) [reply zng/record.go]: deleted

**Pull request merged**